File: lib/domain/cultivation_plan/dtos/public_plan_save_pesticide_reference_row.rb
# frozen_string_literal: true

module Domain
  module CultivationPlan
    module Dtos
      class PublicPlanSavePesticideReferenceRow
        attr_reader :reference_pesticide_id, :reference_crop_id, :reference_pest_id,
                    :name, :active_ingredient, :description, :region,
                    :usage_constraint, :application_detail

        def initialize(
          reference_pesticide_id:,
          reference_crop_id:,
          reference_pest_id:,
          name:,
          active_ingredient: nil,
          description: nil,
          region: nil,
          usage_constraint: nil,
          application_detail: nil
        )
          @reference_pesticide_id = reference_pesticide_id.to_i
          @reference_crop_id = reference_crop_id.to_i
          @reference_pest_id = reference_pest_id.to_i
          @name = name.nil? ? nil : name.to_s
          @active_ingredient = active_ingredient
          @description = description
          @region = region
          @usage_constraint = usage_constraint
          @application_detail = application_detail
          freeze
        end
      end
    end
  end
end

File: crates/agrr-domain/src/cultivation_plan/dtos/public_plan_save_pesticide_reference_row.rs
//! Ruby: `Domain::CultivationPlan::Dtos::PublicPlanSavePesticideReferenceRow`

use super::public_plan_save_pesticide_application_detail_row::PublicPlanSavePesticideApplicationDetailRow;
use super::public_plan_save_pesticide_usage_constraint_row::PublicPlanSavePesticideUsageConstraintRow;

/// Ruby: `Domain::CultivationPlan::Dtos::PublicPlanSavePesticideReferenceRow`
#[derive(Debug, Clone)]
pub struct PublicPlanSavePesticideReferenceRow {
    pub reference_pesticide_id: i64,
    pub reference_crop_id: i64,
    pub reference_pest_id: i64,
    pub name: Option<String>,
    pub active_ingredient: Option<String>,
    pub description: Option<String>,
    pub region: Option<String>,
    pub usage_constraint: Option<PublicPlanSavePesticideUsageConstraintRow>,
    pub application_detail: Option<PublicPlanSavePesticideApplicationDetailRow>,
}

impl PublicPlanSavePesticideReferenceRow {
    pub fn new(
        reference_pesticide_id: i64,
        reference_crop_id: i64,
        reference_pest_id: i64,
        name: Option<String>,
        active_ingredient: Option<String>,
        description: Option<String>,
        region: Option<String>,
        usage_constraint: Option<PublicPlanSavePesticideUsageConstraintRow>,
        application_detail: Option<PublicPlanSavePesticideApplicationDetailRow>,
    ) -> Self {
        Self {
            reference_pesticide_id,
            reference_crop_id,
            reference_pest_id,
            name,
            active_ingredient,
            description,
            region,
            usage_constraint,
            application_detail,
        }
    }
}
